docs/adrs/006-web-apis.md

```diff
@@ -0,0 +1,240 @@
+
+In Penumbra, any blockchain query requires a connection with the wallet provider, therefore all actions and
+queries can be called 'private'. It is reasonable to construct the notion of a **client** – the interface 
+that manages the injected connections and provides methods for interacting with the blockchain.
+
+Creating the `client` should be the starting point for any application working with Penumbra:
+
+```ts
+import { createPenumbraClient } from '@penumbra-zone/client';
+
+export const client = createPenumbraClient();
+```
+
+The flow of work with the `client` would be as follows:
+
+- User creates an instance of the `client`
+- They use the `client` to establish the injected connection, and the `client` saves the connection in its state
+- To fetch information from the wallet, the application passes the `client` instance to the fetch methods
+
+The idea of a single shared `client` is inspired by these popular TypeScript libraries:
+- [Prisma ORM](https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/instantiate-prisma-client)
+- [Supabase](https://supabase.com/docs/reference/javascript/initializing)
+- [Viem](https://viem.sh/docs/clients/intro)
+
+## Injected connection
+
+The `client` must be able to connect to injected wallets. To make the API possible, the interface injected into the `window` object must be standardized:
+
+```ts
+export interface PenumbraProvider extends Readonly<PenumbraStateEventTarget> {
+  /** Should contain a URI at the provider's origin, 
+   * serving a manifest describing this provider */
+  readonly manifest: string;
+
+  /** Call to gain approval. Returns `MessagePort` to this provider. 
+   * Might throw descriptive errors if user denies the connection */
+  readonly connect: () => Promise<MessagePort>;
+
+  /** A synchronous call to retrive the `MessagePort` if the connection was already established */
+  readonly reconnect: () => MessagePort | undefined;
+
+  /** Call to indicate the provider should discard approval of this origin. 
+   * Successfull if doesn't throw errors */
+  readonly disconnect: () => Promise<void>;
+
+  /** Should synchronously return the present connection state.
+   * - `true` indicates active connection.
+   * - `false` indicates connection is closed or rejected.
+   * - `undefined` no attempt has resolved. Connection may be attempted
+   */
+  readonly isConnected: () => boolean | undefined;
+
+  /** Synchronously return present injection state */
+  readonly state: () => PenumbraState;
+
+  /**
+   * A connection listener function. Fires a callback with 
+   * the `origin`, `state`, and `connected` fields each time the state changes
+   */
+  readonly onConnectionChange: (cb: (connection: { origin: string, connected: boolean, state: PenumbraState }) => void) => void;
+}
+```
```

Comment:
It's basically a simplified version of the listeners with added `connected` boolean to the callback. I feel like the `event` in `addEventListener` is redundant since users will only access the `detail` property of `event`. But if we don't expose CustomEvent, then probably it shouldn't be called `addEventListener`. 

We can vote on this design or come up with another version

Reply:
the event interface is standard, and extensible. flexibility may help avoid breaking changes to this api. it may eventually be useful for the provider to emit additional detail with the events, or emit other event types we haven't thought of yet.

for example we could add a specific 'penumbraconnected' event without changing any interface signature in a way that breaks consumers.

use the standard functionality also prevents any implementation errors on our part - all of the callback logic is handled very well by browser developers. we only write the message types.

Reply:
Ok, I agree to compromise. Put back the `addEventListener` and `removeEventListener` but simplified the types definition. In the current implementation (probably why I was so hesitant to accept this API before), CustomEvents aren't really typed, so in Nextjs example I wrote the following:

```ts
const typedEvent = event as CustomEvent<{ origin: string, state: PenumbraState }>;
```

If we fix the types, it'll already be better. Even though I was really tempted to change the `addEventListener` to simply `addListener: (type: 'penumbrastate', cb: (value: { origin: string; connected: boolean; state: PenumbraState }) => void)`